src/main/kotlin/org/jmailen/gradle/kotlinter/support/SortedThreadSafeReporterWrapper.kt

```diff
@@ -0,0 +1,58 @@
+package org.jmailen.gradle.kotlinter.support
+
+import com.github.shyiko.ktlint.core.LintError
+import com.github.shyiko.ktlint.core.Reporter
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ConcurrentMap
+
+/**
+ * A wrapper for a Reporter that guarantees thread safety and consistent ordering of all the calls to the reporter.
+ * As a downside, the calls to the wrapped reporter are delayed until the end of the execution.
+ */
+class SortedThreadSafeReporterWrapper(
```

Comment:
I think this is a fine approach given the data will be small. Could we verify it is needed? When I did the quick look at ktlint, the reporters were thread safe, but I didn't test if result order was stable.

If not, then yeah I think this is reasonable.

Reply:
My analysis:
- https://github.com/pinterest/ktlint/blob/master/ktlint-reporter-checkstyle/src/main/kotlin/com/github/shyiko/ktlint/reporter/checkstyle/CheckStyleReporter.kt : ThreadSafe & sorted
- https://github.com/pinterest/ktlint/blob/3c01b494497798672f714951652df84c4a8e3324/ktlint-reporter-plain/src/main/kotlin/com/github/shyiko/ktlint/reporter/plain/PlainReporter.kt : ThreadSafe **but not sorted**!
- https://github.com/pinterest/ktlint/blob/3c01b494497798672f714951652df84c4a8e3324/ktlint-reporter-json/src/main/kotlin/com/github/shyiko/ktlint/reporter/json/JsonReporter.kt : ThreadSafe & sorted

=> Right now, the thread safety is guaranteed by the current implementations but the consistency of the output is not given in `PlainReporter`. Therefore, I would keep it.